### PR TITLE
Expose a more generic DiffAlgorithm property in CompareOptions

### PR DESCRIPTION
**Pull request by nulltoken**
Following #1039 and in order to avoid future breaking changes maybe would it be wise to expose a property accepting an enum of Diff algorithms (Meyer, Minimal, ...).

/cc @dmalikov 

- Add new property
- Obsolete current property
- Implement minimal tests to cover each algorithm

### CODE REVIEW

**Comment by nulltoken** [comment]
Ready for review

File: LibGit2Sharp/CompareOptions.cs [hunk 1]
```diff
@@ -39,6 +42,12 @@ public CompareOptions()
         /// <summary>
         /// Use the "patience diff" algorithm.
         /// </summary>
+        [Obsolete("This property will be removed in the next release. Please use Algorithm instead.")]
         public bool UsePatienceAlgorithm { get; set; }
+
+        /// <summary>
+        /// Algorithm to be used when performing a Diff.
```
Comment (Therzok) [inline LibGit2Sharp/CompareOptions.cs]: `(Default = Algorithm.Meyers)`

File: LibGit2Sharp/Diff.cs [hunk 2]
```diff
@@ -49,7 +49,7 @@ private static GitDiffOptions BuildOptions(DiffModifiers diffOptions, FilePath[]
                 options.Flags |= GitDiffOptionFlags.GIT_DIFF_INCLUDE_UNMODIFIED;
             }
 
-            if (compareOptions.UsePatienceAlgorithm)
+            if (compareOptions.Algorithm == DiffAlgorithm.Patience)
```
Comment (Therzok) [inline LibGit2Sharp/Diff.cs]: Missing check for DiffAlgorithm.Minimal

File: LibGit2Sharp/DiffAlgorithm.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    /// <summary>
+    /// Algorithm used when performing a Diff.
+    /// </summary>
+    public enum DiffAlgorithm
+    {
+        /// <summary>
+        /// The basic greedy diff algorithm.
+        /// </summary>
+        Meyers = 0,
+
```
Comment (Therzok) [inline LibGit2Sharp/DiffAlgorithm.cs]: Missing DIffAlgorithm.Minimal.

**Comment by Therzok** [comment]
Other than missing implementation for minimal diff, that's all I can say.

**Comment by nulltoken** [comment]
@Therzok I've been refraining from implementing the --minimal` algo as I haven't found a proper concise example of its output (which should differ from what would be computed through Meyer or Patience).

Any idea?

File: LibGit2Sharp/CompareOptions.cs (hunk 1 above)
Comment (nulltoken) [inline LibGit2Sharp/CompareOptions.cs]: Fixed

**Comment by Therzok** [comment]
Only seems to do things for big diffs. I just did:
```
➜  monodevelop git:(master) git diff --diff-algorithm=minimal HEAD~600 HEAD | wc -l
   77496
➜  monodevelop git:(master) git diff --diff-algorithm=myers HEAD~600 HEAD | wc -l
   79049
```

Returned same results until around ~600 commits, and I tracked down what changed the stuff. It seems to be XML that's the culprit of making minimal better, but only in a big diff context.

Here's the diff between the two.

https://gist.github.com/Therzok/0f1f127ab2c9bb95681b

They make chrome choke when trying to gist them:
https://www.dropbox.com/s/dsmizfjupayh6sm/diff-min.diff?dl=0
https://www.dropbox.com/s/rh7kfxo5b9w4fwo/diff-myers.diff?dl=0

Differences also happen in libgit2sharp repo with ~300 commits from HEAD.
https://gist.github.com/Therzok/8c8a87c4586d428513b8

**Comment by Therzok** [comment]
In short it only affects huge diffs. This would be useful for someone diffing HEAD to 1 year prior, but not many people do this. So it would be best to keep it.

Short things short: minimal is just myers with better diffs for huge stuff.

**Comment by nulltoken** [comment]
Proposal: Merge this and log an `up-for-grabs` issue for Minimal implementation. Eventually @dmalikov's **[question](http://permalink.gmane.org/gmane.comp.version-control.git/268852)** may get an answer...

**Comment by nulltoken** [comment]
Published as NuGet pre-release package `LibGit2Sharp.0.22.0-pre20150527180327`

**Pull request merged**